src/System.Linq/tests/WhereTests.cs

```diff
@@ -0,0 +1,644 @@
+
+            Assert.Equal(0, result.Count());
+        }
+
+        [Fact]
+        public void Where_EmptyEnumerable_ReturnsNoElements()
+        {
+            IEnumerable<int> source = Enumerable.Empty<int>();
+            bool wasSelectorCalled = false;
+
+            IEnumerable<int> result = source.Where(value => { wasSelectorCalled = true; return true; });
+            
+            Assert.Equal(0, result.Count());
+            Assert.False(wasSelectorCalled);
+        }
+        
+        [Fact]
+        public void Where_Array_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            int[] source = new[] { 1 };
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+        
+        [Fact]
+        public void Where_List_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            List<int> source = new List<int>() { 1 };
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void Where_IReadOnlyCollection_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            IReadOnlyCollection<int> source = new ReadOnlyCollection<int>(new List<int>() { 1 });
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void Where_ICollection_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            ICollection<int> source = new LinkedList<int>(new List<int>() { 1 });
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void Where_IEnumerable_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            IEnumerable<int> source = Enumerable.Repeat(1, 1);
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void WhereWhere_Array_ReturnsExpectedValues()
```

Comment:
Added tests for WhereSelect, SelectWhere (though these may need to move to the Select tests), and the overload.